pkg/controller/direct/compute/forwardingrule_controller.go

```diff
@@ -0,0 +1,496 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compute
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"k8s.io/klog/v2"
+
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+
+	gcp "cloud.google.com/go/compute/apiv1"
+	computepb "cloud.google.com/go/compute/apiv1/computepb"
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/compute/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const ctrlName = "forwardingrule-controller"
+
+func init() {
+	registry.RegisterModel(krm.ComputeForwardingRuleGVK, NewForwardingRuleModel)
+}
+
+func NewForwardingRuleModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &forwardingRuleModel{config: config}, nil
+}
+
+type forwardingRuleModel struct {
+	config *config.ControllerConfig
+}
+
+// model implements the Model interface.
+var _ directbase.Model = &forwardingRuleModel{}
+
+type forwardingRuleAdapter struct {
+	id                          *ForwardingRuleIdentity
+	forwardingRulesClient       *gcp.ForwardingRulesClient
+	globalForwardingRulesClient *gcp.GlobalForwardingRulesClient
+	desired                     *krm.ComputeForwardingRule
+	actual                      *computepb.ForwardingRule
+}
+
+var _ directbase.Adapter = &forwardingRuleAdapter{}
+
+func (m *forwardingRuleModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.ComputeForwardingRule{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Get ResourceID
+	resourceID, err := refs.GetResourceID(u)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get projectID
+	projectID, err := refs.ResolveProjectID(ctx, reader, u)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get network
+	if obj.Spec.NetworkRef != nil {
+		networkRef, err := ResolveComputeNetwork(ctx, reader, obj, obj.Spec.NetworkRef)
+		if err != nil {
+			return nil, err
+
+		}
+		obj.Spec.NetworkRef.External = networkRef.External
```

Comment:
Do you mean cleaning up other fields like `Name` in `obj.Spec.NetworkRef`? I won't worry about other fields because we get value from External only, see the mapping function: [code](https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/e82cf9bb1aa018e2797eb5faeec634fb2fdd549d/pkg/controller/direct/compute/mapper.go#L88)

Reply:
Right, it might be clearer/safer if you did clear all those other fields, it's a nit because I agree it shouldn't change functionality.